### PR TITLE
BUG: Fix SpatialOrientationAdapter::FromDirectionCosines

### PR DESCRIPTION
**Pull request by blowekamp**
The old algorithm could generate an invalid combinations of axes. The new algorithm is greedy by selecting the element in the matrix with the greatest magnitude, and eleminating invalid elements for further selection.

## PR Checklist
- [ ] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [ ] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)
- [ ] Added test (or behavior not changed)
- [ ] Updated API documentation (or API not changed)
- [ ] Added [license](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Utilities/KWStyle/ITKHeader.h) to new files (if any)
- [ ] Added Python wrapping to new files (if any) as described in [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) Section 9.5
- [ ] Added [ITK examples](https://github.com/InsightSoftwareConsortium/ITKSphinxExamples) for all new major features (if any)

Refer to the [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) for
further development details if necessary.

### CODE REVIEW

File: Modules/Core/Common/src/itkSpatialOrientationAdapter.cxx [hunk 1]
```diff
@@ -32,47 +32,70 @@ namespace itk
 SpatialOrientationAdapter::OrientationType
 SpatialOrientationAdapter::FromDirectionCosines(const DirectionType & Dir)
 {
```
Comment (dzenanz) [inline Modules/Core/Common/src/itkSpatialOrientationAdapter.cxx:34]: Commit message: The old algorithm could generate an invalid combinations of axes.

**Review by dzenanz — APPROVED** [review]
Thanks for adding the fix here too!

**Comment by issakomi** [comment]
Interesting, just discovered strange part in _itkSpatialOrientationAdapter.cxx_  (not part of the PR)

```
// all terms must be defined, otherwise just punt
  if (terms[0] == SpatialOrientation::ITK_COORDINATE_UNKNOWN ||
      terms[1] == SpatialOrientation::ITK_COORDINATE_UNKNOWN || terms[2] == SpatialOrientation::ITK_COORDINATE_UNKNOWN)
  {
    return SpatialOrientation::ITK_COORDINATE_ORIENTATION_RIP;
  }
```
Why RIP? Why not  _ITK_COORDINATE_ORIENTATION_INVALID_ or may be RAI (fallback to identity matrix).
However it is _not_ a request to change this, the file is very critical, smallest issue may cause huge compatibility problems, personally i prefer to stay away.

**Comment by blowekamp** [comment]
> Why RIP? Why not _ITK_COORDINATE_ORIENTATION_INVALID_ or may be RAI (fallback to identity matrix).
> However it is _not_ a request to change this, the file is very critical, smallest issue may cause huge compatibility problems, personally i prefer to stay away.

Yes, I think the intention was to fallback to the default with an identity matrix. For the failure case, that was added in the test, I two axis were given the same label. So it was a failure case not caught by the fallback. This could be changed to verify that the results a valid, but I don't see an easy way to test for the failure case in the bug.

However, with the new algorithm I do not see how unknown, or an invalid combination could be produced. Even with an invalid matrix of all zeros, it should produce a valid results.  So these lines might be able to be removed?

**Pull request merged**